### PR TITLE
Add support to tapyrus-setup sign and createsig commands to sign xfield

### PR DESCRIPTION
**Pull request by Naviabheeman**
This PR adds the command line support to generate a  federation signature to make an xfield change. 

```
-  tapyrus-setup blocks
-  tapyrus-setup sign
-  tapyrus-setup createsig
```

 commands are used to create the xfield signature
 
Use XFieldHash type hash the serialized xfield to generate the signature
Add tests to verify signing xfield instead of block. A valid signature was generated using the commands in a 3 signer, 2 threshold setup and is added to the unit test.

### CODE REVIEW

**Review by azuchi — CHANGES_REQUESTED** [review]

File: tests/resources/federations.toml [hunk 1]
```diff
@@ -1,21 +1,19 @@
 [[federation]]
```
[the remark this answers is not in the document]
Comment (Naviabheeman) [reply tests/resources/federations.toml:1]: This PR is trying to address only tapyrus-signer changes along with the minimum change from other parts required to compile successfully. Sorting and other error handling are all in PR 170. The federation is sorted in the order of block height and validated fully.  The verification key and other unchanged parameters are updated in all federations first. This way all the existing code in block creation and signing can work without breaking even when new xfield is added.

File: src/federation.rs [hunk 2]
```diff
@@ -102,8 +106,13 @@ pub struct Federation {
     /// Verifiable Secret Share and commitments from all signers in the federation.
     /// This field must be None when the signer is not a member of the federation.
     nodevss: Option<Vec<Vss>>,
-    /// The aggregated public key
-    aggregated_public_key: PublicKey,
+    /// The federation parameter
+    /// aggregated public key / max block size
+    xfield: XField,
+    // pubkey for signature verification
+    pub verification_key: Option<PublicKey>,
```
[the remark this answers is not in the document]
Comment (Naviabheeman) [reply src/federation.rs:113]: Yes, verification key is the valid aggregated public key in a block at block_height. The complete code is in PR 170.

File: tests/resources/federations.toml [hunk 3]
```diff
@@ -1,21 +1,19 @@
...
 block-height = 0
-threshold = 3
-aggregated-public-key = "030d856ac9f5871c3785a2d76e3a5d9eca6fcce70f4de63339671dfb9d1f33edb0"
+threshold = 2
```
[the remark this answers is not in the document]
Comment (Naviabheeman) [reply tests/resources/federations.toml:3]: This is added in a new test `test_load_federations_threshold_change` too.

File: src/cli/setup/compute_sig.rs [hunk 4]
```diff
@@ -144,35 +186,64 @@ impl<'a> ComputeSigCommand {
...
         for (sig, public_key) in keyed_local_sigs {
             signatures.insert(SignerID { pubkey: public_key }, (sig.gamma_i, sig.e));
         }
+
+        let hash = if let Some(ref block) = block {
+            block.header.signature_hash().to_vec()
+        } else if xfield != XField::None {
+            xfield.signature_hash().unwrap().to_vec()
+        } else {
+            return Err(Error::InvalidArgs(
```
Comment (Yamaguchi) [inline src/cli/setup/compute_sig.rs:221]: Will this line be executed?
I think this line is unreachable, as the same check is on line 204.

**Review by Yamaguchi — CHANGES_REQUESTED** [review]

File: src/federation.rs [hunk 5]
```diff
@@ -364,15 +512,39 @@ mod tests {
...
+            XField::AggregatePublicKey(TEST_KEYS.aggregated()),
+            None,
+            Some(sig),
         )
     }
 
+    fn valid_federation_maxblocksize() -> Federation {
+        let mut federation = valid_federation();
+
+        federation.xfield = XField::MaxBlockSize(400000);
```
Comment (Yamaguchi) [inline src/federation.rs:544]: Are there any restrictions on the maximum or minimum MaxBlockSize?
If a small value (maybe 170bytes) is specified, will TapyrusCore reject the block?

According to https://github.com/chaintope/tapyrus-core/blob/master/doc/tapyrus/signedblocks.md, the minimum Block Header is 170byte.

**Review by Yamaguchi — CHANGES_REQUESTED** [review]

File: src/cli/setup/compute_sig.rs (hunk 4 above)
Comment (Naviabheeman) [reply src/cli/setup/compute_sig.rs:221]: yes. this is unreachable. I'll fix it.

File: src/federation.rs (hunk 5 above)
Comment (Naviabheeman) [reply src/federation.rs:544]: > Are there any restrictions on the maximum or minimum MaxBlockSize?

Tapyrus Core requires the minimum block size to be 1000 bytes. There is no maximum limit other than the 4 byte capacity allowed.

>If a small value (maybe 170bytes) is specified, will TapyrusCore reject the block?

Yes, Tapyrus Core does not allow max block size in xfield to be set below 1000 bytes. Also, BlockAssembler in Tapyrus Core uses the same minimum limit. However if a smaller size valid block is created outside of Tapyrus Core like using tools like our test framework, there is no size restriction. The block is accepted by the node after validation.

**Review by azuchi — APPROVED** [review]

**Review by Yamaguchi — APPROVED** [review]

**Pull request merged**